Pull in the updates from the mix_bit_upgrades branch

tking53 opened the pull request:
This branch includes upgrades stemming from the work testing the mixed bit same frequency capabilities of utkscan(or).

This branch has been in use with multiple analyses, and are considered stable enough for merging into dev.

Highlights include:
Protections for the poll2 command "toggle". Previously if you gave an invalid bit name (polarity, gain, traces, etc.) toggle would flip bit 0. This behavior has been fixed by having the name -> bit list check fail and output an error before the call to the API.

We also added some safety checks to toggle, adjust_offsets, pwrite and pmwrite. For whatever reason the API lacks sanity checks for these functions and will let you try to run them on a module that is not installed. (a 3 module system would let you write to module 3 rather than just 0,1,2) this would cause the computer to Hard lock and require a power button forced reboot. pread and pmread had protections in the API but i have added our checks as well for consistency's sake. 

We have added support for the 42159 and 42950 ( the newest vandle firmwares 12b, and 16b respectively) 
We have also added support for parsing the 500MHz general firmware. Right now this is Hard matched since we only have 1. This will be fixed once we have more firmwares, and the parsing functions have been updated to reflect the differences in the list more format (specifically Header Word 2 which is structured differently since the "forced trigger bit is now 3 bits )

We have made Ctrl+c clear the line in poll2, rather than just spitting out an error saying we blocked the keystroke 
We have also fixed the INS key behavior, previously you could enter "insert"/"overwrite" mode but were unable to exit it. 

We have given the CfdAnalyzer some of the upgrades that were previously given to the Fitting and Waveform analyzers. Namly the ignore list.

We fixed the MapNodeXmlParser's parsing of the Cfd Parameters, and we have made the parser ALWAYs parse the cfd node for the "vandle" type, which is also done for the Fit and Trace nodes.

Finally the GeProcessor has gotten root output support. It is filled into the clover_vec_  with cloverNum = -1, since detNum is type:subtype specific. This will let us use the two processors together and identify them later in the root analysis.

tking53 committed:
Add protections to invalid bits for `toggle`

previously if you gave a "name" that wasnt in the toggle_names list it would evaluate to 0 and then toggle the 0th bit. This unwanted behavior.

Now we check that if the "bit" argument is a number and < 24 since we can give direct CCSRA bit numbers to `toggle`. If we detect a non-numeric character then we check if the argument is in the valid `toggle_names` list, if its not we return ERROR invalid name/bit.

tking53 committed:
poll2 help documentation update

Updated poll2 help message, mostly reordering the help so that certain commands are next to each other 

And i changed the default `dump` file name to not look like a save output

Changed the failed save help text to remind people to include the ".set" in the file name

tking53 committed:
Changes to extreme baseline variation

Changing the extreme baseline threshold to 15% of the baseline value, We are doing this because 16b can have a variation of 300-400 and still be ok. 

Im also adding an error statement to this, so that for every 10,000 traces we throw out here; we print a note into the terminal. 

Also adding a cut for the avg baseline also being <10 adc units (for every bit resolution) because if the base line is this low then something is wrong. (example: beta traces for RIKEN after implant)

tking53 committed:
Changing Ctrl+c in poll2 to clear the cmd line

This commit changes  the use of <Ctrl>+c in poll2.
Previously all we did was catch the <ctrl>+c and print an error. Now we are using this key stroke to clear the current line in the poll2 prompt. It does not store the partial command in the history.

Currently we are just wrapping the private function Terminal::clear_() with a publicly accessible version, but we should probably reimplement the clearing action with the public nature in-mind

 This has a limited use case but is helpful on occasion.

tking53 committed:
Updating firmware support 

We now fully support the 250MHz firmware received in  2019, namely 42159 and 42950 which are the newest 12b and 16b vandle firmwares respectively

We also have initial support for the 500MHz general firmware. Its hard matched right now while we are confirming everything, but it should have the same list mode as the 34688->42950 series, with the exception of the changes to header word 2 relevant to the 500MHz (changes to the onboard cfd data). If everything checks out we will remove the hard match, and the separate enum value

tking53 committed:
added default value to the global frequency parsing

This sets a default of frequency == 0 which is consistent with the other sanity checks elsewhere in the code.

tking53 committed:
CTerminal toggle-able INS mode 

Previously you could not disable INS mode once it was enabled (without restarting the applications) Now we can

tking53 committed:
Changed ValidFitAnalysis to ValidTimingAnalysis

Generalized the HasValidFitAnalysis stuff to HasValidTimingAnalysis as we move to include the cfd analzyer (poly)

tking53 committed:
Moving the analyzer ignore check to the base class

removed the check against the ignoredTypes_ lists from the individual analyzers and implemented at common version. I also removed the `SetPhase(0.0)` from the saturated/pileup/ignored check, because we don't want to accidentally overwrite the phase during later analyzers (so we are order independent other than the WaveformAnalyzer being first). I moved the SetPhase(0.0) to after all analyzers have been called, so that if it really should be 0 it is (i.e. the timing algorithms failed for some reason).

Added IsIgnoredDetector() which is a precheck for Analyze() this way we don't have to call analyze to check if a detector is in the ignore list. We are doing this so that in the future we can have the polycfd on some detector types and the fitting analyzer on others.

tking53 committed:
Added ignored list to CfdAnalyzer

Beginning to modernize the CfdAnalyzer, starting with the check against an ignore list and that the waveform analyzer completed successfully

tking53 committed:
Fixed parsing of cfd pars from xml

We were passing the <Cfd> node but were expecting the full <Channel> node.

tking53 committed:
Made the fitting and Cfd analyzers compatible

I have added non-destructive checks to see if the timing for a given channel has been done before. this will prevent us from doing the timing twice and possibly overriding a good value with a bad one (if the second analyzer converges on a garbage answer because a parameter was set incorrectly)

tking53 committed:
Moving IsIgnored() to TraceAnalzyer.cpp 

Moving the function for IsIgnored() from the TraceAnalyzer.hpp to the TraceAnalyzer.cpp. Which is the better place for it

tking53 committed:
Updates to DefaultConfigValues && always parse CFD node for vandles

tking53 committed:
New Pol1 method for poly cfd

We now have a 1st order poly nominal for the poly cfd; as well as the ability to switch between the 1st and 2nd order (note this is for the leading "fit") This choice is selectable from the XML and defaults to the linear

Fixed a bug in the CfdAnalyzer, where were were passing CFD:D and CFD:L rather than CFD:F and CFD:D (although F is the only parameter we need for the poly cfd).

tking53 committed:
Fix writing to invalid module

For what ever reason the protections in the API do no apply to pwrite, pmwrite, toggle, and adjust_offsets. We now check that the module entered on the command line is a "valid" module, this check is dependent on the Pixie Cards being loaded sequentially into the slots. ( We always do this anyways). I have added the same protection to pread and pmread for consistency's sake. This should be all of the commands that have support for the ranges.

tking53 committed:
Changed numeric check to "built-in" version

A numeric check already existed in the code, so i have changed the check in toggle for a valid ccsra bit to use this function rather than its own.

tking53 committed:
Adding "onlydoubles" flag to VandleProcessor

This xml flag will fill the root equivalent of the vandle only hists. this way we can do things like check the gain matching etc (things without a start detector) but it is important that this flag is disabled (false) when a trigger is present or you will double count

tking53 committed:
Reverting VandleProcessor OnlyDoubles

causes double/triple counting. implementation needs to be fixed

tking53 committed:
Add Root Output to the GeProcessor

This reuses the Clover Root structure; It sets the cloverNum to -1 so that we can tell them apart if they were ever used together.  

This Compiles but needs actual testing. (Using git to pass the code around due to COVID19)

rin-yokoyama commented on Analysis/Utkscan/analyzers/source/WaveformAnalyzer.cpp:
```diff
@@ -64,14 +73,26 @@ void WaveformAnalyzer::Analyze(Trace &trace, const ChannelConfiguration &cfg) {
+        //Also if the avg baseline is lower than 10 ADC units (for any bit resolution),
+        // this is also a sign of a bad trace capture.
+        
+        static const double extremeBaselineVariation = 0.15 * baseline.first; //Checking for an std of 15% of the avg baseline (this way we are sensitive to the different bit resolutions)
```
We should remove static keyword here. Otherwise the extremeBaselineVariation is fixed to the value in the first call which I think is not intended.

tking53 replied:
yes your right.

tking53 replied:
your right.

tking53 committed:
Waveform Analyzer :Removing unwanted static keyword

Forgot to remove the leftover static keyword for the extremeBaselineVariation.

Thanks to @rin-yokoyama for catching that

rin-yokoyama reviewed (approved):
It now looks good to merge
